### PR TITLE
fix linter findings

### PR DESCRIPTION
**Pull request by ibihim**
I ran golangci-lint on the repo and fixed most of the issues. On others I left a comment for discussion.

Signed-off-by: Krzysztof Ostrowski <kostrows@redhat.com>

### CODE REVIEW

File: pkg/config/stringsource.go [hunk 1]
```diff
@@ -60,6 +60,9 @@ func ResolveStringValue(s configv1.StringSource) (string, error) {
 		return "", fmt.Errorf("no valid PEM block of type %q found in key", StringSourceKeyBlockType)
 	}
 
+	// Deprecated: Legacy PEM encryption as specified in RFC 1423 is insecure by
```
Comment (stlaz) [inline pkg/config/stringsource.go:63]: so what's the alternative?

File: pkg/groupmapper/groupmapper_test.go [hunk 2]
```diff
@@ -86,7 +86,7 @@ func TestUserGroupsMapper_UserFor(t *testing.T) {
 			fakeGroupsClient := fakeuserclient.NewSimpleClientset(groupObjs...)
 
 			userInformer := userinformer.NewSharedInformerFactory(fakeGroupsClient, 5*time.Second)
-			userInformer.User().V1().Groups().Informer().AddIndexers(cache.Indexers{
+			_ = userInformer.User().V1().Groups().Informer().AddIndexers(cache.Indexers{
```
Comment (stlaz) [inline pkg/groupmapper/groupmapper_test.go:89]: `require.NoError(...`

File: pkg/oauth/handlers/default_auth_handler_test.go [hunk 3]
```diff
@@ -182,10 +182,12 @@ func TestWithChallengeErrorsAndMergedSuccess(t *testing.T) {
 	if !handled {
 		t.Error("Expected handling.")
 	}
-	if !(reflect.DeepEqual(map[string][]string(responseRecorder.HeaderMap), expectedHeader1) || reflect.DeepEqual(map[string][]string(responseRecorder.HeaderMap), expectedHeader2)) {
-		t.Errorf("Expected %#v or %#v, got %#v.", expectedHeader1, expectedHeader2, responseRecorder.HeaderMap)
+
+	rec := responseRecorder.Result()
```
Comment (stlaz) [inline pkg/oauth/handlers/default_auth_handler_test.go:186]: what's rec? Use a variable name that is self-explanatory

File: pkg/oauth/handlers/default_auth_handler_test.go [hunk 4]
```diff
@@ -229,19 +231,21 @@ func TestWithRedirect(t *testing.T) {
...
 	if err != nil {
 		t.Errorf("Unexpected error: %v", err)
 	}
 	if !handled {
 		t.Error("Expected handling.")
 	}
-	if !reflect.DeepEqual(map[string][]string(responseRecorder.HeaderMap), expectedHeader1) {
-		t.Errorf("Expected %#v, got %#v.", expectedHeader1, responseRecorder.HeaderMap)
+
+	rec := responseRecorder.Result()
```
Comment (stlaz) [inline pkg/oauth/handlers/default_auth_handler_test.go:243]: ditto

File: pkg/oauthserver/oauth_apiserver_test.go [hunk 5]
```diff
@@ -34,7 +34,7 @@ func TestGetInvalidSessionSecretsFile(t *testing.T) {
 	}
 	defer os.Remove(tmpfile.Name())
 
-	ioutil.WriteFile(tmpfile.Name(), []byte("invalid content"), os.FileMode(0600))
+	_ = ioutil.WriteFile(tmpfile.Name(), []byte("invalid content"), os.FileMode(0600))
```
Comment (stlaz) [inline pkg/oauthserver/oauth_apiserver_test.go:37]: `require.NoError(t, ...`

File: pkg/oauthserver/oauth_apiserver_test.go [hunk 6]
```diff
@@ -57,7 +57,7 @@ func TestGetEmptySessionSecretsFile(t *testing.T) {
 	if err != nil {
 		t.Errorf("Unexpected error: %v", err)
 	}
-	ioutil.WriteFile(tmpfile.Name(), []byte(yaml), os.FileMode(0600))
+	_ = ioutil.WriteFile(tmpfile.Name(), []byte(yaml), os.FileMode(0600))
```
Comment (stlaz) [inline pkg/oauthserver/oauth_apiserver_test.go:60]: `require.NoError(t, ...`

File: pkg/oauthserver/oauth_apiserver_test.go [hunk 7]
```diff
@@ -84,7 +84,7 @@ func TestGetValidSessionSecretsFile(t *testing.T) {
...
```
Comment (stlaz) [inline pkg/oauthserver/oauth_apiserver_test.go:87]: `require.NoError(t, ...`

**Review by stlaz — CHANGES_REQUESTED** [review]
/lgtm cancel

We can easily treat the errors in the tests so I propose to just do that.

File: pkg/config/stringsource.go (hunk 1 above)
Comment (ibihim) [reply pkg/config/stringsource.go:63]: I don't know, as we are just the receiver of that unsecure thingy :smile:

File: pkg/oauth/handlers/default_auth_handler_test.go (hunk 3 above)
Comment (ibihim) [reply pkg/oauth/handlers/default_auth_handler_test.go:186]: It is quite common to use short variable names in Go and rec being a common short name for a recorder.

https://github.com/golang/go/wiki/CodeReviewComments#variable-names=

File: pkg/oauth/external/handler.go [hunk 8]
```diff
@@ -211,12 +211,12 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
 
 func (h *Handler) handleError(err error, w http.ResponseWriter, req *http.Request) {
-	handled, err := h.errorHandler.AuthenticationError(err, w, req)
+	handled, _ := h.errorHandler.AuthenticationError(err, w, req)
 	if handled {
 		return
 	}
 	w.WriteHeader(http.StatusInternalServerError)
-	w.Write([]byte(`An error occurred`))
+	_, _ = w.Write([]byte(`An error occurred`))
```
Comment (stlaz) [inline pkg/oauth/external/handler.go:219]: better yet: use `http.Error(...)` https://pkg.go.dev/net/http#Error

File: pkg/oauth/handlers/default_auth_handler.go [hunk 9]
```diff
@@ -43,10 +43,8 @@ const (
 	// WarningHeaderOpenShiftSource is the name of the agent adding the warning header
 	WarningHeaderOpenShiftSource = "Origin"
 
-	warningHeaderCodeIndex  = 1
-	warningHeaderAgentIndex = 2
-	warningHeaderTextIndex  = 3
-	warningHeaderDateIndex  = 4
+	warningHeaderCodeIndex = 1
```
Comment (stlaz) [inline pkg/oauth/handlers/default_auth_handler.go:46]: eh... maybe keep the old indices in a comment?

File: pkg/server/grant/grant.go [hunk 10]
```diff
@@ -195,7 +195,7 @@ func (l *Grant) handleGrant(user user.Info, w http.ResponseWriter, req *http.Req
 		return
 	}
 
-	req.ParseForm()
```
Comment (stlaz) [inline pkg/server/grant/grant.go:198]: wouldn't it make sense to handle the error here?

File: pkg/oauth/handlers/default_auth_handler_test.go (hunk 3 above)
Comment (ibihim) [reply pkg/oauth/handlers/default_auth_handler_test.go:186]: oh, yeah. this is true. :man_facepalming:

File: pkg/oauth/handlers/default_auth_handler.go (hunk 9 above)
Comment (ibihim) [reply pkg/oauth/handlers/default_auth_handler.go:46]: hm, dead code is bad code. I would rather suggest the person to go through the git commit history. WDYT?

File: pkg/server/grant/grant.go (hunk 10 above)
Comment (ibihim) [reply pkg/server/grant/grant.go:198]: sorry, I just blindly implemented the obvious, not any handling logic, let me do so.

Comment (ibihim) [reply pkg/server/grant/grant.go:198]: @stlaz, in `handleGrant` we don't return error http status codes at all, should I add them there? :sweat_smile:

File: pkg/config/stringsource.go (hunk 1 above)
Comment (ibihim) [reply pkg/config/stringsource.go:63]: Let me add a TODO upfront, so we can tackle it with the TODO story, that I just created.

File: pkg/oauth/handlers/default_auth_handler.go (hunk 9 above)
Comment (stlaz) [reply pkg/oauth/handlers/default_auth_handler.go:46]: dead code indeed is bad, yet seeing indices that seem randomly chosen is puzzling and will raise more eyebrows. I'd rather we keep the unused consts commented :neutral_face:

Comment (ibihim) [reply pkg/oauth/handlers/default_auth_handler.go:47]: ok, I added a comment.

Comment (ibihim) [reply pkg/oauth/handlers/default_auth_handler.go:47]:
```
// There were more indexes, but where unused.
```

File: pkg/server/login/login.go [hunk 11]
```diff
@@ -189,7 +189,7 @@ func (l *Login) handleLogin(w http.ResponseWriter, req *http.Request) {
 
 	audit.AddDecisionAnnotation(req, audit.AllowDecision)
 	klog.V(4).Infof(`Login with provider %q succeeded for %q: %#v`, l.provider, username, authResponse.User)
-	ok, err = l.auth.AuthenticationSucceeded(authResponse.User, then, w, req)
```
Comment (stlaz) [inline pkg/server/login/login.go]: this code should stay

**Review by stlaz — CHANGES_REQUESTED** [review]

File: pkg/server/login/login.go [hunk 12]
```diff
@@ -190,8 +189,16 @@ func (l *Login) handleLogin(w http.ResponseWriter, req *http.Request) {
 
 	audit.AddDecisionAnnotation(req, audit.AllowDecision)
 	klog.V(4).Infof(`Login with provider %q succeeded for %q: %#v`, l.provider, username, authResponse.User)
-	l.auth.AuthenticationSucceeded(authResponse.User, then, w, req)
-	metrics.RecordFormPasswordAuth(metrics.SuccessResult)
```
Comment (stlaz) [inline pkg/server/login/login.go:194]: looks like the success result disappeared?

File: pkg/cmd/oauth-server/cmd.go [hunk 13]
```diff
@@ -63,8 +63,14 @@ func NewOsinServerCommand(out, errout io.Writer, stopCh <-chan struct{}) *cobra.
 	options.Audit.AddFlags(flags)
 
 	flags.StringVar(&options.ConfigFile, "config", "", "Location of the osin configuration file to run from.")
-	cmd.MarkFlagFilename("config", "yaml", "yml")
-	cmd.MarkFlagRequired("config")
+	if err := cmd.MarkFlagFilename("config", "yaml", "yml"); err != nil {
+		fmt.Fprintf(errout, "read config option: %v", err)
```
Comment (stlaz) [inline pkg/cmd/oauth-server/cmd.go:67]: I don't understand what this error message is telling me as a user

File: pkg/cmd/oauth-server/cmd.go [hunk 14]
```diff
@@ -63,8 +63,14 @@ func NewOsinServerCommand(out, errout io.Writer, stopCh <-chan struct{}) *cobra.
...
+		os.Exit(254)
+	}
+	if err := cmd.MarkFlagRequired("config"); err != nil {
+		fmt.Fprintf(errout, "mark config as requied: %v", err)
```
Comment (stlaz) [inline pkg/cmd/oauth-server/cmd.go:71]: I don't understand what this error message is telling me as a user

File: pkg/cmd/oauth-server/cmd.go [hunk 15]
```diff
@@ -63,8 +63,14 @@ func NewOsinServerCommand(out, errout io.Writer, stopCh <-chan struct{}) *cobra.
 	options.Audit.AddFlags(flags)
 
 	flags.StringVar(&options.ConfigFile, "config", "", "Location of the osin configuration file to run from.")
-	cmd.MarkFlagFilename("config", "yaml", "yml")
-	cmd.MarkFlagRequired("config")
+	if err := cmd.MarkFlagFilename("config", "yaml", "yml"); err != nil {
+		fmt.Fprintf(errout, "read config option: %v", err)
+		os.Exit(254)
```
Comment (stlaz) [inline pkg/cmd/oauth-server/cmd.go:68]: No `os.Exit` in library functions! The only acceptable `os.Exit` here is the one in the command's `Run` callback.

File: pkg/oauth/external/handler.go [hunk 16]
```diff
@@ -211,12 +211,14 @@ func (h *Handler) ServeHTTP(w http.ResponseWriter, req *http.Request) {
...
-	handled, err := h.errorHandler.AuthenticationError(err, w, req)
+	handled, _ := h.errorHandler.AuthenticationError(err, w, req)
 	if handled {
 		return
 	}
+
+	klog.V(4).Infof("handle error failed for err: %v", err)
 	w.WriteHeader(http.StatusInternalServerError)
-	w.Write([]byte(`An error occurred`))
+	http.Error(w, "An error occured", http.StatusInternalServerError)
```
Comment (stlaz) [inline pkg/oauth/external/handler.go:221]: You'll want to remove the `w.WriteHeader` above - https://cs.opensource.google/go/go/+/refs/tags/go1.18.1:src/net/http/server.go;l=2093

File: pkg/userregistry/identitymapper/strategy_add.go [hunk 17]
```diff
@@ -33,7 +34,12 @@ func (s *StrategyAdd) UserForNewIdentity(ctx context.Context, preferredUserName
 		desiredUser := &userapi.User{}
 		desiredUser.Name = preferredUserName
 		desiredUser.Identities = []string{identity.Name}
-		s.initializer.InitializeUser(identity, desiredUser)
+
+		err = s.initializer.InitializeUser(identity, desiredUser)
```
Comment (stlaz) [inline pkg/userregistry/identitymapper/strategy_add.go:38]: This error is unused later so you can just paste it into the condition. Also, don't rewrite the err from the wrapping block

File: pkg/userregistry/identitymapper/strategy_add.go [hunk 18]
```diff
@@ -33,7 +34,12 @@ func (s *StrategyAdd) UserForNewIdentity(ctx context.Context, preferredUserName
...
+		if err != nil {
+			return nil, fmt.Errorf("initialize user with identity (%v): %w", identity, err)
```
Comment (stlaz) [inline pkg/userregistry/identitymapper/strategy_add.go:40]: "failed to initialize ..."

File: pkg/userregistry/identitymapper/strategy_add.go [hunk 19]
```diff
@@ -46,8 +52,12 @@ func (s *StrategyAdd) UserForNewIdentity(ctx context.Context, preferredUserName
 		persistedUser.Identities = append(persistedUser.Identities, identity.Name)
 		// If our newly added identity is the only one, initialize the user
 		if len(persistedUser.Identities) == 1 {
-			s.initializer.InitializeUser(identity, persistedUser)
+			err = s.initializer.InitializeUser(identity, persistedUser)
+			if err != nil {
+				return nil, fmt.Errorf("initialize user with identity (%v): %w", identity, err)
```
Comment (stlaz) [inline pkg/userregistry/identitymapper/strategy_add.go:57]: "failed to initialize ..."

File: pkg/userregistry/identitymapper/strategy_add.go [hunk 20]
```diff
@@ -46,8 +52,12 @@ func (s *StrategyAdd) UserForNewIdentity(ctx context.Context, preferredUserName
 		persistedUser.Identities = append(persistedUser.Identities, identity.Name)
 		// If our newly added identity is the only one, initialize the user
 		if len(persistedUser.Identities) == 1 {
-			s.initializer.InitializeUser(identity, persistedUser)
+			err = s.initializer.InitializeUser(identity, persistedUser)
```
Comment (stlaz) [inline pkg/userregistry/identitymapper/strategy_add.go:55]: don't rewrite the err from the wrapping block

File: pkg/userregistry/identitymapper/strategy_claim.go [hunk 21]
```diff
@@ -51,7 +51,12 @@ func (s *StrategyClaim) UserForNewIdentity(ctx context.Context, preferredUserNam
 		desiredUser := &userapi.User{}
 		desiredUser.Name = preferredUserName
 		desiredUser.Identities = []string{identity.Name}
-		s.initializer.InitializeUser(identity, desiredUser)
+
+		err = s.initializer.InitializeUser(identity, desiredUser)
```
Comment (stlaz) [inline pkg/userregistry/identitymapper/strategy_claim.go:55]: don't rewrite the err from the wrapping block

File: pkg/userregistry/identitymapper/strategy_claim.go [hunk 22]
```diff
@@ -63,7 +68,11 @@ func (s *StrategyClaim) UserForNewIdentity(ctx context.Context, preferredUserNam
 		// If this user has no other identities, claim, initialize, and update
 		if len(persistedUser.Identities) == 0 {
 			persistedUser.Identities = []string{identity.Name}
-			s.initializer.InitializeUser(identity, persistedUser)
+			err = s.initializer.InitializeUser(identity, persistedUser)
```
Comment (stlaz) [inline pkg/userregistry/identitymapper/strategy_claim.go:71]: don't rewrite the err from the wrapping block

File: pkg/userregistry/identitymapper/strategy_generate.go [hunk 23]
```diff
@@ -63,7 +63,12 @@ UserSearch:
 			desiredUser := &userapi.User{}
 			desiredUser.Name = potentialUserName
 			desiredUser.Identities = []string{identity.Name}
-			s.initializer.InitializeUser(identity, desiredUser)
+
+			err = s.initializer.InitializeUser(identity, desiredUser)
```
Comment (stlaz) [inline pkg/userregistry/identitymapper/strategy_generate.go:67]: don't rewrite the err from the wrapping block

File: pkg/cmd/oauth-server/cmd.go (hunk 15 above)
Comment (ibihim) [reply pkg/cmd/oauth-server/cmd.go:68]: > os.Exit(255)

https://github.com/openshift/oauth-server/pull/98/files#diff-520d203d74db4a7e2864c3aacc58dc6f70d6f802fda4245334ab0ecd71244d89R53

> klog.Fatal(err) // => os.Exit(255)

https://github.com/openshift/oauth-server/pull/98/files#diff-520d203d74db4a7e2864c3aacc58dc6f70d6f802fda4245334ab0ecd71244d89R41

Could you elaborate, why this is fine here?

**Pull request merged**